server/plugin.go

```diff
@@ -45,8 +52,35 @@ type Plugin struct {
 
 func (p *Plugin) OnActivate() error {
 	p.logger = &metricsLogger{api: p.API}
-
 	appCfg := p.API.GetConfig()
+
+	p.closeChan = make(chan bool)
+	p.waitGroup = sync.WaitGroup{}
+	
+	// we are using a mutually exclusive lock to run a single instance of this plugin
+	// we don't really need to collect metrics twice, in any case we can still collect
+	// twice, TSDB will takce care of overlapped blocks. But it will increase the disk
+	// writes to the remote or local disk.
+	if p.isHA() {
+		var err error
+		p.singletonLock, err = cluster.NewMutex(p.API, "metrics")
+		if err != nil {
+			return err
+		}
+
+		// the constant '20' is determined by healthcheck of the plugin which is 30 seconds.
+		ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
```

Comment:
I don't get this comment. Do we just need something smaller than 30 seconds?

Reply:
Yes, otherwise the server will try to restart the plugin. Since the lock refresh is 15 seconds, 20 is a good time to give up.